mbuild/formats/hoomdxml.py

```diff
@@ -134,95 +135,99 @@ def write_hoomdxml(structure, filename, forcefield, box, ref_distance=1.0,
+                bond_types = np.zeros(len(bonds), dtype=int)
             else:
-                unique_bond_types = dict(enumerate(set([(round(bond.type.k,3),
-                                                         round(bond.type.req,3)) for bond in structure.bonds])))
-                unique_bond_types = {y:x for x,y in unique_bond_types.items()}
-                bond_types = [unique_bond_types[(round(bond.type.k,3),
-                                                 round(bond.type.req,3))] for bond in structure.bonds]
+                unique_bond_types = dict(enumerate(set([(round(bond.type.k, 3),
+                                                         round(bond.type.req, 3)) for bond in structure.bonds])))
+                unique_bond_types = {y: x for x, y in unique_bond_types.items()}
+                bond_types = [unique_bond_types[(round(bond.type.k, 3),
+                                                 round(bond.type.req, 3))] for bond in structure.bonds]
                 xml_file.write('<bond_coeffs>\n')
-                for params,idx in unique_bond_types.items():
-                    xml_file.write('{}\t{}\t{}\n'.format(idx,((params[0]*2.)/ref_energy)*(ref_distance**2.),params[1]/ref_distance))
+                for params, idx in unique_bond_types.items():
+                    xml_file.write('{}\t{}\t{}\n'.format(idx, ((params[0]*2.) / ref_energy)*(ref_distance**2.), params[1]/ref_distance))
                 xml_file.write('</bond_coeffs>\n')
             xml_file.write('<bond>\n')
-            for idx,bond in enumerate(bonds):
-                xml_file.write('{}\t{}\t{}\n'.format(bond_types[idx],*bond))
+            for idx, bond in enumerate(bonds):
+                xml_file.write('{}\t{}\t{}\n'.format(bond_types[idx], *bond))
             xml_file.write('</bond>\n')
 
-        angles = [[angle.atom1.idx, 
-                   angle.atom2.idx, 
+        angles = [[angle.atom1.idx,
+                   angle.atom2.idx,
                    angle.atom3.idx] for angle in structure.angles]
         if angles:
-            unique_angle_types = dict(enumerate(set([(round(angle.type.k,3), 
-                                                      round(angle.type.theteq,3)) for angle in structure.angles])))
-            unique_angle_types = {y:x for x,y in unique_angle_types.items()}
-            angle_types = [unique_angle_types[(round(angle.type.k,3), 
-                                               round(angle.type.theteq,3))] for angle in structure.angles]
+            unique_angle_types = dict(enumerate(set([(round(angle.type.k, 3),
+                                                      round(angle.type.theteq, 3)) for angle in structure.angles])))
+            unique_angle_types = {y: x for x, y in unique_angle_types.items()}
+            angle_types = [unique_angle_types[(round(angle.type.k, 3),
+                                               round(angle.type.theteq, 3))] for angle in structure.angles]
```

Comment:
Yeah the code to get the unique bond/angle/dihedral types is definitely really messy, and I'm sure there's a better way to do this.  I can look into the change you're suggesting @ctk3b, but this is something I tried when putting the writer together originally and ran into some issues. Specifically `unique_angle_types` would feature duplicate types (i.e. the same k and theta values) if the three atom types defining the angle were different.  E.g. angle_type1 between atoms of type 'A', 'A', 'A' was considered different from angle_type2 between atoms of type 'A', 'A', 'B' even if the k and theta values for angle_type1 and angle_type2 were equivalent.